Updated Construction section added Authentication section:

tjlytle opened the pull request:
@leggetter @timcraft  This should address what needs to be done with JWTs. Let me know if anything seems missing.

tjlytle committed:
Updated construction added authentication section:
- Made oauth support optional.
- Addressed JWT generation.
- Limited clients to *one* credential per type, yet supporting multiple types.
- Changes should maintain a 1:1 replationship between an account an a client instance.

leggetter commented on SPECIFICATION.md:
```diff
@@ -41,13 +41,28 @@ Specification
-- Must support API credentials (key and secret) for authentication.
-- Must support OAuth1 token for authentication.
-- Must require an authentication method (credentials or oauth) for construction.
+- Must support multiple authentication credentials (see '[]Authenticating Requests](authenticating-requests)')
```
malformed markdown link

leggetter commented on SPECIFICATION.md:
```diff
@@ -41,13 +41,28 @@ Specification
+
+- Must allow a combination (one or more) of the following credential types:
+    - Must allow an API Key and Secret (current API)
+    - Must allow a signature secret (use for _some_ API requests, and to validate WebHook signatures)
```
Could we considering this being a "Should"?

tjlytle replied:
Specifically the signature? That works for me.

leggetter replied:
Yeah, the "signature secret". It's not really required.

leggetter commented on SPECIFICATION.md:
```diff
@@ -41,13 +41,28 @@ Specification
+- Must allow a combination (one or more) of the following credential types:
+    - Must allow an API Key and Secret (current API)
+    - Must allow a signature secret (use for _some_ API requests, and to validate WebHook signatures)
+    - Must allow multiple private keys for JWT authentication (replacing key and secret)
```
How would you therefore know which one to use when making a request?

tjlytle replied:
Client library should know based on the request. New Voice API needs JWT, current application API key and secret, etc.

leggetter replied:
I can see how the JWT vs. key/secret selection can be made. The question is referring to when there are "multiple private keys" how do you know which of the private keys should be used?

tjlytle replied:
That's the line below that wasn't clear, client library should not accept multiple private keys. If you need that you're likely working with multiple applications, and can then just use multiple instances of the library. Goal here is to eventually reduce to a private key. This complexity is during the transition only.

tjlytle replied:
And total miss on my part, had a few drafts considering the complexity, should be:

> Must allow a private key

Not multiple keys.

leggetter commented on SPECIFICATION.md:
```diff
@@ -41,13 +41,28 @@ Specification
+    - Must allow an API Key and Secret (current API)
+    - Must allow a signature secret (use for _some_ API requests, and to validate WebHook signatures)
+    - Must allow multiple private keys for JWT authentication (replacing key and secret)
+    - May allow OAuth1 token for authentication
```
Could this be a "Should"?

tjlytle replied:
Added as `may` because not sure how JWTs will deprecate OAUth support. Can change to should.

leggetter commented on SPECIFICATION.md:
```diff
@@ -41,13 +41,28 @@ Specification
+    - Must allow a signature secret (use for _some_ API requests, and to validate WebHook signatures)
+    - Must allow multiple private keys for JWT authentication (replacing key and secret)
+    - May allow OAuth1 token for authentication
+- Must not allow more than one credential per type.
```
Could this please be clarified? I'm not sure what it means.

tjlytle replied:
Sure, point here is while you could create a client instance with both API key / secret _and_ a private key, you shouldn't be able to create one with two private keys. 

Would this work?

> Must not allow multiple credentials of the same type (two private keys, two sets of key and secret, etc).

leggetter commented on SPECIFICATION.md:
```diff
@@ -41,13 +41,28 @@ Specification
+- Must determine what type of authentication credential should be used for a request.
+- May allow choosing a type of authentication credential when an API supports more than one.
+- May generate a short lived JWT per request.
+- May generate a longer loved JWT for multiple requests.
```
"loved" -> "lived"

tjlytle replied:
Hey, maybe I love my JWT.

leggetter commented on SPECIFICATION.md:
```diff
@@ -41,13 +41,28 @@ Specification
+- May generate a longer loved JWT for multiple requests.
+- Must provide a method to generate a JWT:
+    - Must allow user provided application id.
+    - Must allow optional user provided valid timestamps (`nbf` and `exp`)
```
Could we please clarify what these abbreviations mean?

tjlytle committed:
A few typos and clarifications:
- Fixed contridiction on multiple private keys
- Expanded JWT timstamps to be more descriptive.

leggetter commented on SPECIFICATION.md:
```diff
@@ -51,17 +51,18 @@ The current state of the API allows multiple authentication credentials. The cli
-    - May allow OAuth1 token for authentication
-- Must not allow more than one credential per type.
+    - Should allow a signature secret (use for _some_ API requests, and to validate WebHook signatures)
+    - Must allow a private key for JWT authentication (replacing key and secret)
```
Should this be stated in a similar way to signature secret? e.g. 

Must allow a private key for JWT authentication (required for _some_ API requests)

tjlytle replied:
Perhaps, how about:

> Must allow a private key for JWT authentication (eventually replacing key and secret)

Point was that JWT should be more and more used, as it replaces other auths. Relates to @timcraft's note below.

leggetter commented on SPECIFICATION.md:
```diff
@@ -51,17 +51,18 @@ The current state of the API allows multiple authentication credentials. The cli
-    - Must allow multiple private keys for JWT authentication (replacing key and secret)
-    - May allow OAuth1 token for authentication
-- Must not allow more than one credential per type.
+    - Should allow a signature secret (use for _some_ API requests, and to validate WebHook signatures)
```
"use for _some_ API requests"

use -> used

?

tjlytle replied:
Who broke my keyboard with all these typos.

timcraft commented:
How does "Must determine what type of authentication credential should be used for a request" work if an API supports multiple credential types?

JWT is presumably the preferred type, where supported?

For older APIs that support key+secret or signature secret, which should be the preference?

tjlytle commented:
@timcraft In that case, would be (which I should add to the spec):
- JWT
- OAuth
- Signature Secret
- Key / Secret

timcraft commented:
@tjlytle Might be useful to document more clearly which APIs support which credentials somewhere? Even if the libraries are coded to select from the correct credentials for a particular method call, developers still need to know which credentials to provide when creating a client.

tjlytle commented:
@timcraft Good point. @leggetter @iain-cox thoughts on that?

> Might be useful to document more clearly which APIs support which credentials somewhere?

tjlytle committed:
Clarified usage of different authentication types.

leggetter commented:
> Might be useful to document more clearly which APIs support which credentials somewhere?

Great point @timcraft :+1:

@iain-cox - do you think that's something we could add to the appropriate sections? At minimum the API reference sections maybe?

iain-cox commented:
yup

tjlytle committed:
Update for file path question.

tjlytle commented:
@leggetter @sammachin @cbetta @judy2k Added like about strings / file paths. Is it clear enough? Goal is to make no assumptions of how a developer will store the credentials. I feel like 

> as string values

may make it seem like a hash / array of strings (say for OAuth) wouldn't be acceptable. But that should be, just don't want to force the developer to store the keys in the filesystem. Is there a clearer way to address this? Any objections?

cbetta commented:
I think this is clear enough.

tjlytle committed:
Allowing user defined parameters:
- Allows JWT body to have additional parameters.
- Bettr defines default values.
